### PR TITLE
feat: Mobile notifications support

### LINKED ISSUE #2170: mobile: create notification handler interface
**What I'm trying to do**

Create an interface that will handle the notification of events. This interface will be implemented by objects within mobile apps to take advantage of [being able to pass target language objects to Go](https://godoc.org/golang.org/x/mobile/cmd/gobind#hdr-Passing_target_language_objects_to_Go).

Update 

**Expected result**

The changes should allow mobile apps to

- Subscribe to a topic.
- Create implementations of handlers that can be passed into the configuration of an Aries mobile agent.

**Sample code**

```go
// Golang
type Handler interface {
    Handle(topic string, payload []byte) error
}
```

gomobile will generate Java/Objective-C code that will contain an interface named Handler.
Mobile apps can create an implementation of it like so:

``` java
// Java
public class MyHandler extends AriesApi.Handler {
    public void handle(String topic, byte[] message) {
        // do something here
    }
}
```

```objc
// Objective-C
@interface MyHandler: NSObject<ApiHandler>{

}
@end

@implementation MyHandler

- (BOOL) handle: (NSString *)topic message:(NSData *)message
          error:(NSError * _Nullable __autoreleasing *)error {
        // do something here
}

@end
```

And then it'll pass an instance of `MyHandler` to Go by calling a new function called `AriesAgent.Registerhandler`.
```go
// rest implementation will use WebSockets
// command implementation will use a Notifier (github.com/hyperledger/aries-framework-go/pkg/controller/webnotifier)
func (a *Aries) RegisterHandler(h Handler, topic string, payload []byte) string {
    /* ... */
}
```
```java
// Java
            // register handler
            Handler handler = new MyHandler();
            String registrationID = ariesAgent.registerHandler(handler, "didexchange_states");
```
```objc
// Objective-C

    // register handler
    MyHandler *handler = [[MyHandler alloc] init];
    NSString *regID = [ariesAgent registerHandler:handler topics:@"didexchange_states"];
```

**Questions and Concerns**

- [x] What about remote agents i.e. the agents running on external servers that mobile apps can connect to using our SDK
~~- [ ] Lifecycle: TODO (when exactly does the notifier get used internally?)~~
- [x] Unregistering notifiers:
  - [X] By who (the mobile app or the SDK)? -> mobile app

### PR DESCRIPTION
**Pull request by soluchok**
Adds mobile notifications support.
* works on iOS
* works on Android (haven’t checked yet)

The main idea is to use the `Handler` interface with the `Handle` method which plays the role of a callback function.
Usage: The client needs to implement the `Handler` interface. Notifications will be sent through this callback function (`Handle`).

Closes #2170

Signed-off-by: Andrii Soluk <isoluchok@gmail.com>

### CODE REVIEW

File: cmd/aries-agent-mobile/pkg/api/notifier.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package api
```
[the remark this answers is not in the document]
Comment (ToJen) [reply cmd/aries-agent-mobile/pkg/api/notifier.go:1]: and we should also add a comment at the top that the file is not expected to be used by the mobile app and so gomobile will not compile all the code in it

**Review by ToJen — CHANGES_REQUESTED** [review]

**Review by ToJen — APPROVED** [review]
looks good, we just have this extra file: test/bdd/fixtures/agent-rest/couchdb-config/10-single-node.ini

**Comment by soluchok** [comment]
> looks good, we just have this extra file: test/bdd/fixtures/agent-rest/couchdb-config/10-single-node.ini

It was done by make command. I didn't want to create a separate PR for it. So, I just included it in this PR.

File: cmd/aries-agent-mobile/pkg/wrappers/command/aries.go [hunk 2]
```diff
@@ -149,6 +171,56 @@ func populateHandlers(commands []command.Handler, pkgMap map[string]map[string]c
...
+				}
+			}
+		}
+		a.mutex.RUnlock()
+	}
+}
+
+// RegisterHandler registers a handler to process incoming notifications from the framework.
+// Handler is implemented by mobile apps.
+func (a *Aries) RegisterHandler(h api.Handler, topics string) string {
```
Comment (llorllale) [inline cmd/aries-agent-mobile/pkg/wrappers/command/aries.go:193]: `topics` can be a var-arg

File: cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go [hunk 3]
```diff
@@ -42,7 +54,121 @@ func NewAries(opts *config.Options) (*Aries, error) {
...
+func (ar *Aries) startNotificationListener() { // nolint: gocyclo,funlen
+	if ar.WebsocketURL == "" {
+		return
+	}
+
+	conn, _, err := websocket.Dial(context.Background(), ar.WebsocketURL, &websocket.DialOptions{ //nolint:bodyclose
+		HTTPClient: &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true, // nolint: gosec
```
Comment (llorllale) [inline cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go:86]: why is this enabled?

File: cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go [hunk 4]
```diff
@@ -42,7 +54,121 @@ func NewAries(opts *config.Options) (*Aries, error) {
...
+
+			// exit if websocket is closed
+			if websocket.CloseStatus(err) != -1 {
+				return
+			}
+
+			continue
+		}
+
+		ar.mutex.RLock()
```
Comment (llorllale) [inline cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go:120]: should defer `RUnlock` here

**Comment by llorllale** [comment]
@soluchok 

> > looks good, we just have this extra file: test/bdd/fixtures/agent-rest/couchdb-config/10-single-node.ini
> 
> It was done by make command. I didn't want to create a separate PR for it. So, I just included it in this PR.

Please remove this file from this PR. PR #2180 had a bug that was later fixed in #2182.

**Review by llorllale — CHANGES_REQUESTED** [review]
Please remove couchdb config file.

Also see my other comments.

File: cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go (hunk 4 above)
Comment (soluchok) [reply cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go:120]: it is a loop. `defer ` will not work

**Comment by soluchok** [comment]
> Please remove couchdb config file.
> 
> Also see my other comments.

@llorllale Removed `couchdb` config file.

File: cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go (hunk 3 above)
Comment (soluchok) [reply cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go]: Fixed!

File: cmd/aries-agent-mobile/pkg/wrappers/command/aries.go (hunk 2 above)
Comment (soluchok) [reply cmd/aries-agent-mobile/pkg/wrappers/command/aries.go:193]: @llorllale As far as I remember we cannot use variadic arguments.  @ToJen Could you please confirm this?
e.g RegisterHandler(h api.Handler, topics ...string) is not possible to use, @ToJen correct?.

File: cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go (hunk 4 above)
Comment (llorllale) [reply cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go:111]: `defer ar.mutex.RUnlock()` will not work because what follows is a loop?

Comment (soluchok) [reply cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go]: it fixed

**Review by llorllale — APPROVED** [review]

**Comment by llorllale** [comment]
@soluchok 

> > looks good, we just have this extra file: test/bdd/fixtures/agent-rest/couchdb-config/10-single-node.ini
> 
> It was done by make command. I didn't want to create a separate PR for it. So, I just included it in this PR.

You're right about this file - the couchdb image is setting its access mode to read-only. You can include it in this PR.

File: cmd/aries-agent-mobile/pkg/wrappers/command/aries.go (hunk 2 above)
Comment (ToJen) [reply cmd/aries-agent-mobile/pkg/wrappers/command/aries.go:193]: yes we can only use simple args like `string`

**Review by rolsonquadras — APPROVED** [review]
It would be good if we can document the usage of notifiers in [REAME](https://github.com/hyperledger/aries-framework-go/tree/master/cmd/aries-agent-mobile#3-usage) file

**Comment by ToJen** [comment]
@rolsonquadras i will do it in a separate commit

**Pull request merged**